### PR TITLE
Backend integration

### PR DESCRIPTION
**Pull request by AniiMancuso**
## Summary ##

**Trello card**: https://trello.com/c/T6MZcs2c

Details of the books almost achieved, missing bug fix tab bar

### iPhone SE

Captura de Pantalla 2022-10-14 a la(s) 15 09 04
Captura de Pantalla 2022-10-14 a la(s) 15 09 09

### iPhone 14 Pro

Captura de Pantalla 2022-10-14 a la(s) 15 09 25
Captura de Pantalla 2022-10-14 a la(s) 15 09 28
Captura de Pantalla 2022-10-14 a la(s) 15 09 32

## Known issues ##

Fix in the navigation bar, we are seeing why the tab bar is seen in the detail view of the books. The alert is already presented and the post request is made correctly

### CODE REVIEW

File: WBooks/BookRepository.swift [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+//  WBooks
+//
+//  Created by ana.mancuso on 30/09/2022.
+//
+import Alamofire
+
+class BookRepository {
+    
+func fetchBooks(onSuccess: @escaping ([Library]) -> Void, onError: @escaping (Error) -> Void) {
+        let url = URL(string: "https://private-deb86-wbooksiostraining.apiary-mock.com/books")!
```
Comment (dfmm472) [inline WBooks/BookRepository.swift:12]: align

File: WBooks/BookRepository.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    AF.request(url, method: .get).responseJSON { response in
+
```
Comment (dfmm472) [inline WBooks/BookRepository.swift]: delete break line

File: WBooks/ImageUrl.swift [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+            guard let url = URL(string: link) else {
+
+                return self.image = UIImage(named: "defaultImage")
+
+            }
+
+            downloaded(from: url, contentMode: mode)
+
+        }
```
Comment (dfmm472) [inline WBooks/ImageUrl.swift]: reformat text

File: WBooks/Library/Details/DetailsViewController.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+class DetailsViewController: UIViewController {
+
+    private lazy var detailsV = DetailsView()
+    private let detailsViewModel: DetailsViewModel
+    
+    init (detailsViewModel: DetailsViewModel) {
+        self.detailsViewModel = detailsViewModel
+        super.init(nibName: nil, bundle: nil)
+    }
```
Comment (dfmm472) [inline WBooks/Library/Details/DetailsViewController.swift:22]: add break line

File: WBooks/Library/Details/DetailsViewController.swift [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            guard let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath) as? ReviewTableViewCell  else {
+                fatalError("The dequeued cell is not an instance of LibraryTableViewCell.")
+            }
+            cell.setupReviewData(reviewer: detailsViewModel.reviewer[indexPath.row])
+            return cell
+        }
+
+        func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
+            return true
+        }
```
Comment (dfmm472) [inline WBooks/Library/Details/DetailsViewController.swift]: Is this code snippet aligned?

File: WBooks/Library/Details/DetailsView.swift [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+//            let detailsViewModel = DetailsViewModel.init(bookDetails: Library)
+//            let detailsViewController = DetailsViewController(detailsViewModel: detailsViewModel)
+//            detailsViewController.goToRentals()
+        } else {
+            alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: "Default action"), style: .default, handler: { _ in
+            NSLog("The \"OK\" alert occured.")
+            }))
+//            self.present(alert, animated: true, completion: nil)
+        }
+    }
```
Comment (dfmm472) [inline WBooks/Library/Details/DetailsView.swift]: the actions of the buttons must go in the controller, if the assignment is made to code if the assignment must be made from the view, but the action must be in the viewController

File: WBooks/Library/Details/DetailsViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    
+    func loadSampleLibrarys() {
+        reviewRepository.fetchBooks { [weak self] comments in
+            guard let selfAux = self else {
+                return
+            }
+            selfAux.reviewer = comments
+            selfAux.changeList?()
+        } onError: { errorLista in
+            print(errorLista.localizedDescription)
```
Comment (dfmm472) [inline WBooks/Library/Details/DetailsViewModel.swift]: change for debugPrint

File: WBooks/Library/Details/Reviewer.swift [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        }
+
+        enum BookKey: String, CodingKey {
+            case id = "id"
+            case name = "name"
+            case image = "image"
+            case comment = "comment"
+        }
+    
+}
```
Comment (dfmm472) [inline WBooks/Library/Details/Reviewer.swift]: Is this code snippet aligned?

File: WBooks/Library/Details/Reviews/ReviewTableViewCell.swift [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        backgroundColor = .clear
+        reviewView.layer.cornerRadius = 10
+        reviewView.layer.shadowColor = UIColor.darkGray.cgColor
+        reviewView.layer.shadowRadius = 3
+        reviewView.layer.shadowOpacity = 0.2
+        reviewView.layer.shadowOffset = CGSize(width: 0.0, height: 0.1)
+        reviewView.layer.masksToBounds = false
```
Comment (dfmm472) [inline WBooks/Library/Details/Reviews/ReviewTableViewCell.swift]: could be added in the didSet of the reviewView

File: WBooks/Library/Library.swift [hunk 10]
```diff
@@ -4,24 +4,41 @@
...
+
+        enum BookKey: String, CodingKey {
+            case id = "id"
+            case title = "title"
+            case author = "author"
+            case genre = "genre"
+            case year = "year"
+            case status = "status"
+            case image = "image"
+        }
```
Comment (dfmm472) [inline WBooks/Library/Library.swift]: Is this code snippet aligned?

File: WBooks/Library/LibraryTableViewModel.swift [hunk 11]
```diff
@@ -9,23 +9,24 @@ import UIKit
...
-        library = [cover1, cover2, cover3, cover4, cover5]
-         
+        bookRepository.fetchBooks { [weak self] listaDeLibros in
+            guard let selfAux = self else {
+                return
+            }
+            selfAux.library = listaDeLibros
+            selfAux.changeList?()
+        } onError: { errorLista in
+            print(errorLista.localizedDescription)
```
Comment (dfmm472) [inline WBooks/Library/LibraryTableViewModel.swift]: change for debugPrint

File: WBooks/PostRepository.swift [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+//  WBooks
+//
+//  Created by ana.mancuso on 12/10/2022.
+//
+import Alamofire
+
+class PostRepository {
+    
+func fetchBooks(onSuccess: @escaping ([Post]) -> Void, onError: @escaping (Error) -> Void) {
+        let url = URL(string: "https://private-deb86-wbooksiostraining.apiary-mock.com/rents")!
```
Comment (dfmm472) [inline WBooks/PostRepository.swift]: align

File: WBooks/PostRepository.swift [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+                guard let rents = try? JSONDecoder().decode([Post].self, from: JSONbooks) else {
+                    onError(rentsError.decodeError)
+                    return
+                }
+
+                onSuccess(rents)
+
+            case .failure(let error):
+                onError(error)
+            }
```
Comment (dfmm472) [inline WBooks/PostRepository.swift]: Is this code snippet aligned?

File: WBooks/Rentals/Post.swift [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+        enum BookKey: String, CodingKey {
+            case id = "id"
+            case title = "title"
+            case author = "author"
+            case genre = "genre"
+            case year = "year"
+            case status = "status"
+            case image = "image"
+        }
```
Comment (dfmm472) [inline WBooks/Rentals/Post.swift]: Is this code snippet aligned?

File: WBooks/Rentals/RentalsViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    
+    func loadPost() {
+        postRepository.fetchBooks { [weak self] post in
+            guard let selfAux = self else {
+                return
+            }
+            selfAux.postBook = post
+            selfAux.changeList?()
+        } onError: { errorLista in
+            print(errorLista.localizedDescription)
```
Comment (dfmm472) [inline WBooks/Rentals/RentalsViewModel.swift]: change for debugPrint

File: WBooks/Rentals/RentalsViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,31 @@
+//
+//  RentalsViewModel.swift
+//  WBooks
+//
+//  Created by ana.mancuso on 12/10/2022.
+//
+import UIKit
```
Comment (dfmm472) [inline WBooks/Rentals/RentalsViewModel.swift]: need to import UIKit?

File: WBooks/Rentals/Post.swift [hunk 17]
```diff
@@ -0,0 +1,44 @@
+//
+//  Post.swift
+//  WBooks
+//
+//  Created by ana.mancuso on 12/10/2022.
+//
+
+import UIKit
```
Comment (dfmm472) [inline WBooks/Rentals/Post.swift]: need to import UIKit?

File: WBooks/Library/Library.swift [hunk 18]
```diff
@@ -4,24 +4,41 @@
 //
 //  Created by ana.mancuso on 20/09/2022.
 //
-
 import UIKit
```
Comment (dfmm472) [inline WBooks/Library/Library.swift]: need to import UIKit?

File: WBooks/Library/Details/Reviewer.swift [hunk 19]
```diff
@@ -0,0 +1,34 @@
+//
+//  Reviewer.swift
+//  WBooks
+//
+//  Created by ana.mancuso on 12/10/2022.
+//
+import UIKit
```
Comment (dfmm472) [inline WBooks/Library/Details/Reviewer.swift]: need to import UIKit?

File: WBooks/Library/Details/Reviewer.swift [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+            photo = try! container.decode(String.self, forKey: .image)
+            comment = try! container.decode(String.self, forKey: .comment)
+        }
+
+        enum BookKey: String, CodingKey {
+            case id = "id"
+            case name = "name"
+            case image = "image"
+            case comment = "comment"
+        }
```
Comment (dfmm472) [inline WBooks/Library/Details/Reviewer.swift]: if the name is equal to the variable there is no need to put this

File: WBooks/Library/Library.swift (hunk 10 above)
Comment (dfmm472) [inline WBooks/Library/Library.swift]: if the name is equal to the variable there is no need to put this

File: WBooks/Rentals/Post.swift (hunk 14 above)
Comment (dfmm472) [inline WBooks/Rentals/Post.swift]: if the name is equal to the variable there is no need to put this

**Review by DiegoHernandez22 — APPROVED** [review]

File: WBooks/AppDelegate.swift [hunk 21]
```diff
@@ -12,8 +12,14 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
-        UINavigationBar.appearance().backgroundColor = UIColor(red: 56/255, green: 170/255, blue: 235/255, alpha: 1)
-        UINavigationBar.appearance().titleTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.white]
+        let appearance = UINavigationBarAppearance()
+        appearance.backgroundColor =  UIColor(red: 56/255, green: 170/255, blue: 235/255, alpha: 1)
+        appearance.titleTextAttributes = [.foregroundColor: UIColor.white]
+        appearance.largeTitleTextAttributes = [.foregroundColor: UIColor.white]
+        UINavigationBar.appearance().tintColor = .white
+        UINavigationBar.appearance().standardAppearance = appearance
+        UINavigationBar.appearance().compactAppearance = appearance
+        UINavigationBar.appearance().scrollEdgeAppearance = appearance
```
Comment (dfmm472) [inline WBooks/AppDelegate.swift:27]: extract in a function

File: WBooks/BookRepository.swift [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+                guard let books = try? JSONDecoder().decode([Library].self, from: JSONbooks) else {
+                    onError(BookError.decodeError)
+                    return
+                }
+                onSuccess(books)
+            case .failure(let error):
+                onError(error)
+            }
+        }
+        }
```
Comment (dfmm472) [inline WBooks/BookRepository.swift]: space format

File: WBooks/BookRepository.swift [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+                guard let comments = try? JSONDecoder().decode([Reviewer].self, from: JSONbooks) else {
+                    onError(BookError.decodeError)
+                    return
+                }
+                onSuccess(comments)
+            case .failure(let error):
+                onError(error)
+            }
+        }
+        }
```
Comment (dfmm472) [inline WBooks/BookRepository.swift]: space format

File: WBooks/ImageUrl.swift [hunk 24]
```diff
@@ -0,0 +1,42 @@
+//
+//  ImageUrl.swift
+//  WBooks
+//
+//  Created by ana.mancuso on 12/10/2022.
+//
+import UIKit
+let imageCache = NSCache<AnyObject, AnyObject>()
```
Comment (dfmm472) [inline WBooks/ImageUrl.swift:8]: add break line

File: WBooks/Library/Cell/LibraryTableViewCell.swift [hunk 25]
```diff
@@ -8,33 +8,37 @@
...
+    
+    @IBOutlet weak var libraryView: UIView! {
+        didSet {
+            libraryView.layer.cornerRadius = 10
+            libraryView.layer.shadowColor = UIColor.darkGray.cgColor
+            libraryView.layer.shadowRadius = 3
+            libraryView.layer.shadowOpacity = 0.2
+            libraryView.layer.shadowOffset = CGSize(width: 0.0, height: 0.1)
+            libraryView.layer.masksToBounds = false
+        }
```
Comment (dfmm472) [inline WBooks/Library/Cell/LibraryTableViewCell.swift:22]: cool 😎

File: WBooks/Login/LoginView.swift [hunk 26]
```diff
@@ -13,9 +13,10 @@ final class LoginView: NibView {
     @IBOutlet weak var passwordTextField: UITextField!
     @IBOutlet weak var iconImage: UIImageView!
     @IBOutlet weak var backgroundImage: UIImageView!
-    @IBOutlet var singinButton: UIButton! {
+    @IBOutlet var signinButton: UIButton! {
         didSet {
-            singinButton.setTitle(NSLocalizedString("SIGN_IN_BUTTON", comment: ""), for: .normal)
+            signinButton.setTitle(NSLocalizedString("SIGN_IN_BUTTON", comment: ""), for: .normal)
+
```
Comment (dfmm472) [inline WBooks/Login/LoginView.swift]: delete break line

**Review by dfmm472 — APPROVED** [review]

**Pull request merged**